packages/sdk/src/universe/builders/single-token/index.ts

```diff
@@ -0,0 +1,126 @@
+import { BaseBuilder } from "../base";
+import { Order } from "../../order";
+import * as Types from "../../types";
+import { BytesEmpty, lc, s } from "../../../utils";
+import { BigNumber, constants, utils } from "ethers/lib/ethers";
+import { REPL_MODE_STRICT } from "repl";
+import { AssetClass } from "../../types";
+
+export class SingleTokenBuilder extends BaseBuilder {
+  public isValid(order: Order): boolean {
+    try {
+      const copyOrder = this.build({
+        ...order.params,
+      });
+
+      if (!copyOrder) {
+        return false;
+      }
+
+      if (copyOrder.hashOrderKey() !== order.hashOrderKey()) {
+        return false;
+      }
+    } catch {
+      return false;
+    }
+
+    return true;
+  }
+
+  public build(params: Types.Order) {
+    this.defaultInitialize(params);
+    //TODO: Add partial filling logic in the future
+
+    return new Order(this.chainId, {
+      kind: "single-token",
+      side: params.side,
+      hash: params.hash,
+      maker: lc(params.maker),
+      make: {
+        assetType: {
+          assetClass: params.make.assetType.assetClass,
+          ...(params.make.assetType.contract && {
+            contract: lc(params.make.assetType.contract),
+          }),
+          ...(params.make.assetType.tokenId && {
+            tokenId: params.make.assetType.tokenId,
+          }),
+        },
+        value: params.make.value,
+      },
+      taker: lc(params.taker),
+      take: {
+        assetType: {
+          assetClass: params.take.assetType.assetClass,
+          ...(params.take.assetType.contract && {
+            contract: lc(params.take.assetType.contract),
+          }),
+          ...(params.take.assetType.tokenId && {
+            tokenId: params.take.assetType.tokenId,
+          }),
+        },
+        value: params.take.value,
+      },
+      salt: params.salt,
+      start: params.start!,
+      end: params.end!,
+      type: params.type,
+      data: params.data,
+      signature: params?.signature,
+      makeBalance: params.makeBalance,
+      makeStock: params.makeStock,
+    });
+  }
+
+  public buildMatching(
+    order: Types.Order,
+    taker: string,
+    data: { amount?: string }
+  ) {
+    const rightOrder = {
+      type: order.type,
+      maker: taker,
+      taker: constants.AddressZero,
+      make: JSON.parse(JSON.stringify(order.take)),
```

Comment:
Why do we need the parse/stringify logic here?

Reply:
This way we dereference the value that's set. Othewise if we do
```
      make: order.take,
      take: order.take,
```

After that if we go into the if case
```
    if (AssetClass.ERC1155 == order.make.assetType.assetClass) {
    ...
    ...
      rightOrder.take.value = Math.floor(Number(data.amount)).toString();
    }

```

`rightOrder.take.value` is a reference to `make.value` --> both `make.value` and `take.value` will be set to that value.